pkg/registry/apis/dashboard/register.go

```diff
@@ -144,42 +183,365 @@ func (b *DashboardsAPIBuilder) InstallSchema(scheme *runtime.Scheme) error {
+	}
+
+	newAccessor, err := utils.MetaAccessor(newDashObj)
+	if err != nil {
+		return fmt.Errorf("error getting meta accessor: %w", err)
+	}
+
+	// Parse namespace for old dashboard
+	nsInfo, err := types.ParseNamespace(oldAccessor.GetNamespace())
+	if err != nil {
+		return fmt.Errorf("failed to parse namespace: %w", err)
+	}
+
+	// Validate requester's organization context
+	id, err := identity.GetRequester(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting requester: %w", err)
+	}
+
+	// Validate organization access
+	if id.GetOrgID() != nsInfo.OrgID {
+		return apierrors.NewNotFound(a.GetResource().GroupResource(), a.GetName())
+	}
+
+	// Basic validations
+	if err := b.validateBasicProperties(title, newAccessor); err != nil {
+		return err
+	}
+
+	// Validate folder existence if specified and changed
+	if newAccessor.GetFolder() != "" && newAccessor.GetFolder() != oldAccessor.GetFolder() {
+		if err := b.validateFolderExists(ctx, newAccessor.GetFolder(), nsInfo.OrgID); err != nil {
+			return err
 		}
 	}
 
+	// Validate refresh interval
+	if err := b.validateRefreshInterval(refresh); err != nil {
+		return err
+	}
+
+	// Check for provisioning - disallow updates to provisioned dashboards if not allowed
+	if err := b.validateProvisionedDashboardUpdate(ctx, oldAccessor); err != nil {
+		return err
+	}
+
+	// Validate generation conflicts
+	if oldAccessor.GetGeneration() > newAccessor.GetGeneration() {
+		return dashboards.ErrDashboardVersionMismatch // TODO: Check if this is the correct error to return
+	}
+
 	return nil
 }
 
+// validateBasicProperties validates basic dashboard properties
+func (b *DashboardsAPIBuilder) validateBasicProperties(title string, accessor utils.GrafanaMetaAccessor) error {
+	// Validate title
+	if title == "" {
+		return fmt.Errorf("dashboard title cannot be empty")
+	}
+
+	if len(title) > 5000 {
+		return fmt.Errorf("dashboard title is too long (max 5000 characters)")
+	}
+
+	// Validate UID - get it from accesso
+	uid := accessor.GetName()
+	if uid != "" {
+		// Check UID length
+		if len(uid) > 40 { // TODO: Set correct length
+			return fmt.Errorf("dashboard UID is too long (max 40 characters)")
+		}
+
+		// Check valid UID format using regex
+		uidPattern := regexp.MustCompile(`^[a-zA-Z0-9\-\_]+$`) // TODO: Make const outside
+		if !uidPattern.MatchString(uid) {
+			return fmt.Errorf("dashboard UID can only contain alphanumeric characters, dashes and underscores")
+		}
+	}
+
+	// Validate message
+	if message := accessor.GetMessage(); message != "" && len(message) > 500 {
+		return fmt.Errorf("dashboard update message is too long (max 500 characters)")
+	}
+
+	return nil
+}
+
+// validateFolderExists checks if a folder exists
+func (b *DashboardsAPIBuilder) validateFolderExists(ctx context.Context, folderUID string, orgID int64) error {
+	// Check if folder exists using the folder store
+	_, err := b.folderClient.Get(ctx, folderUID, orgID, metav1.GetOptions{})
```

Comment:
This looks quite inefficient.

I'm not quite sure how much of a concern it is in reality in terms of performance (though my hunch is that this won't work very well for Booking et al.) so I'd rather have @ryantxu chime in.

I guess short-term this is OK but I wonder if we could delegate this check to Unified Storage if both dashboards and folders are running in appropriate dual-writer modes (i.e. the storage can find both the dashboard & the folder internally).

Reply:
Yeah now that I look at it, that's something I should delegate down to the storage. Thanks!

Reply:
Done